### PR TITLE
allow for selecting configurable subjects on creating printable reports

### PR DESCRIPTION
**Pull request by dsnyder-fairway**
issue was that if you selected a configurable subject, the subject_id would default to null, which would is equivalent to All subjects

### CODE REVIEW

File: report-processor/src/main/resources/application.sql.yml [hunk 1]
```diff
@@ -22,12 +22,13 @@ sql:
         ${sql.snippet.writingTraits},
         ${sql.snippet.exam}
       JOIN district d ON s.district_id=d.id
+      LEFT JOIN subject sub ON sub.id=a.subject_id
```
Comment (wtritch) [inline report-processor/src/main/resources/application.sql.yml]: The `${sql.snippet.exam}` snippet already references the subject table as `su`:
```
    exam: >-
        ${sql.snippet.examColumns},
        st.ssid as student_ssid,
        st.gender_code as student_gender_code,
        st.last_or_surname as student_last_name,
        st.first_name as student_first_name,
        a.id as asmt_id,
        a.name as asmt_name,
        a.label as asmt_label,
        a.grade_code as asmt_grade_code,
        a.type_id as asmt_type_id,
        su.code as asmt_subject_code,
        a.school_year as asmt_school_year,
        a.min_score as asmt_min_score,
        a.max_score as asmt_max_score,
        a.cut_point_1 as asmt_cut_point_1,
        a.cut_point_2 as asmt_cut_point_2,
        a.cut_point_3 as asmt_cut_point_3,
        a.cut_point_4 as asmt_cut_point_4,
        a.cut_point_5 as asmt_cut_point_5
        from exam e
          join student st on st.id = e.student_id
          join asmt a on e.asmt_id = a.id
          join subject su on a.subject_id=su.id
          join school s on e.school_id = s.id
          join school isch on st.inferred_school_id = isch.id
```

Comment (dsnyder-fairway) [reply report-processor/src/main/resources/application.sql.yml]: ah, thank you. I'll update this

File: report-processor/src/main/resources/application.sql.yml [hunk 2]
```diff
@@ -68,6 +70,8 @@ sql:
         wt.performance_task_writing_type,
         ${sql.snippet.exam}
       JOIN district d ON s.district_id=d.id
+      LEFT JOIN subject sub ON sub.id=a.subject_id
+      LEFT JOIN asmt_type asmt_t ON asmt_t.id=a.type_id
```
Comment (wtritch) [inline report-processor/src/main/resources/application.sql.yml]: I'm not sure you want a `LEFT JOIN` here because then the filter may not be applied?  @agorina ?

File: report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/util/ReportProcessorUtils.java [hunk 3]
```diff
@@ -32,14 +18,8 @@
      */
     public static Map<String, Object> getExamFilterParameters(final ExamQueryParams queryParameters) {
         final Map<String, Object> parameters = new HashMap<>();
-        parameters.put("subject_id", queryParameters.getSubjectCode() == null
-                ? null
```
Comment (agorina) [inline report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/util/ReportProcessorUtils.java]: Isn't 'null' subject a valid condition for a student group with the 'null' subject (aka ALL subjects)? Why would configurable subjects change it?

File: report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/util/ReportProcessorUtils.java [hunk 4]
```diff
@@ -32,14 +18,8 @@
...
-                : SubjectIdByCode.get(queryParameters.getSubjectCode())
-        );
-        parameters.put("assessment_type", queryParameters.getAssessmentTypeCode() == null
-                ? null
-                : AssessmentTypeIdByCode.get(queryParameters.getAssessmentTypeCode())
-        );
```
Comment (agorina) [inline report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/util/ReportProcessorUtils.java]: Could you please clarify why this change is needed? This seems to be supporting any subjects?

File: report-processor/src/main/resources/application.sql.yml [hunk 5]
```diff
@@ -27,7 +27,7 @@ sql:
       ${sql.reportProcessor.snippet.percentileLeftJoin}
       WHERE e.student_id in (:student_ids)
         AND (:school_year is null or e.school_year=:school_year)
-        AND (:subject_id is null or a.subject_id=:subject_id)
+        AND (:subject_code is null or su.code=:subject_code)
```
Comment (agorina) [inline report-processor/src/main/resources/application.sql.yml]: Similar to the above comment, I do not understand why the change?

File: report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/util/ReportProcessorUtils.java [hunk 6]
```diff
@@ -32,14 +18,8 @@
      */
     public static Map<String, Object> getExamFilterParameters(final ExamQueryParams queryParameters) {
         final Map<String, Object> parameters = new HashMap<>();
-        parameters.put("subject_id", queryParameters.getSubjectCode() == null
-                ? null
-                : SubjectIdByCode.get(queryParameters.getSubjectCode())
-        );
-        parameters.put("assessment_type", queryParameters.getAssessmentTypeCode() == null
-                ? null
```
Comment (agorina) [inline report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/util/ReportProcessorUtils.java]: Why do we support 'null' asmt type here? Is it to represent 'all' asmt types?

Comment (dsnyder-fairway) [reply report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/util/ReportProcessorUtils.java]: Yeah

File: report-processor/src/main/resources/application.sql.yml [hunk 7]
```diff
@@ -68,6 +68,7 @@ sql:
         wt.performance_task_writing_type,
         ${sql.snippet.exam}
       JOIN district d ON s.district_id=d.id
+      LEFT JOIN asmt_type asmt_t ON asmt_t.id=a.type_id
```
Comment (agorina) [inline report-processor/src/main/resources/application.sql.yml]: Why `LEFT JOIN`?

File: report-processor/src/main/resources/application.sql.yml [hunk 8]
```diff
@@ -81,9 +82,9 @@ sql:
       ) as wt on wt.exam_id = e.id
       WHERE e.student_id in (:student_ids)
         AND (:school_year is null or e.school_year=:school_year)
-        AND (:subject_id is null or a.subject_id=:subject_id)
+        AND (:subject_code is null or su.code=:subject_code)
         AND (:grade_id is null or a.grade_id=:grade_id)
-        AND a.type_id=:type_id
+        AND asmt_t.code=:asmt_type_code
```
Comment (agorina) [inline report-processor/src/main/resources/application.sql.yml]: Could `asmt_type_code` be null?

File: report-processor/src/main/resources/application.sql.yml (hunk 5 above)
Comment (dsnyder-fairway) [reply report-processor/src/main/resources/application.sql.yml]: before we were using the ReportProcessorUtil to retrieve the id from an outdated static mapping. This uses the database to determine the subject code

File: report-processor/src/main/resources/application.sql.yml (hunk 8 above)
Comment (dsnyder-fairway) [reply report-processor/src/main/resources/application.sql.yml]: I didn't think so here, since the type_id could have been null as well

File: report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/util/ReportProcessorUtils.java [hunk 9]
```diff
@@ -1,29 +1,15 @@
...
 
-    private static final Map<String, Integer> SubjectIdByCode = ImmutableMap.of(
-            "Math", 1,
-            "ELA", 2
-    );
-
-    private static final Map<String, Integer> AssessmentTypeIdByCode = ImmutableMap.of(
-            "ica", 1,
-            "iab", 2,
-            "sum", 3
```
Comment (agorina) [inline report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/util/ReportProcessorUtils.java]: I do not think this has changed. Configurable subjects do not change assessment types. It would be cleaner to use AssessmentType from the RDW_Common though.

File: report-processor/src/main/resources/application.sql.yml (hunk 5 above)
Comment (agorina) [reply report-processor/src/main/resources/application.sql.yml]: Got it.

File: report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/util/ReportProcessorUtils.java (hunk 9 above)
Comment (dsnyder-fairway) [reply report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/util/ReportProcessorUtils.java]: Is there any reason why we would not want the asmt types from the database though? I did this to allow for flexibility

Comment (agorina) [reply report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/util/ReportProcessorUtils.java]: Good point, makes sense.

File: report-processor/src/main/resources/application.sql.yml [hunk 10]
```diff
@@ -98,6 +99,7 @@ sql:
         wt.performance_task_writing_type,
         ${sql.snippet.exam}
       JOIN district d ON s.district_id=d.id
+      LEFT JOIN asmt_type asmt_t ON asmt_t.id=a.type_id
```
Comment (agorina) [inline report-processor/src/main/resources/application.sql.yml]: And the same comment about `LEFT JOIN` here?

File: report-processor/src/main/resources/application.sql.yml (hunk 7 above)
Comment (dsnyder-fairway) [reply report-processor/src/main/resources/application.sql.yml]: Good question. I can probably change this to a `JOIN`

File: report-processor/src/main/resources/application.sql.yml [hunk 11]
```diff
@@ -317,6 +319,8 @@ sql:
             JOIN asmt a ON e.asmt_id = a.id
             JOIN school s ON e.school_id = s.id
             JOIN school isch ON st.inferred_school_id = isch.id
+            LEFT JOIN subject sub ON sub.id=a.subject_id
+            LEFT JOIN asmt_type asmt_t ON asmt_t.code=a.type_id
```
Comment (agorina) [inline report-processor/src/main/resources/application.sql.yml]: Ditto about the `LEFT JOIN`.

File: report-processor/src/main/resources/application.sql.yml (hunk 8 above)
Comment (agorina) [reply report-processor/src/main/resources/application.sql.yml]: It feel strange that we have `AND (:asmt_type_code is null or asmt_t.code = :asmt_type_code)` in the code below but not here. 

 But if it was this way and works...(I admit I am not familiar with this query).

File: report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/util/ReportProcessorUtils.java (hunk 9 above)
Comment (dsnyder-fairway) [reply report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/util/ReportProcessorUtils.java]: went ahead and used AssessmentType form RDW_Common. Testing out some other exports while changing the from `LEFT JOIN` to `JOIN` resulted in some errors
